Compression.md

```diff
@@ -0,0 +1,145 @@
+
+WebSocket compression is a WebSocket extension.
+
+### 2. WebSocket Compression: the permessage-deflate algorithm
+Permessage-deflate is a WebSocket extension defined by [RFC7692](https://tools.ietf.org/html/rfc7692) which provides a specification for the compression functionality. It defines the negotiation process and a compression algorithm called `DEFLATE`. 
+
+A `permessage-deflate` negotiation happens during the upgrade request from HTTP to WebSocket. A `permessage-deflate` negotiation offer has a mandatory `permessage-deflate` string followed by a semi-colon separated list of extension parameters. There are four extension parameters defined for WebSocket compression:
+ - `client_no_context_takeover`, `server_no_context_takeover`
+ - `client_max_window_bits`, `server_max_window_bits`
+We shall revisit these parameters in the future sections, where we discuss their use and effects.
+
+A permessage-deflate negotiation response has a mandatory `permessage-deflate` string followed by a semi-colon separated list of extension parameters, agreed upon by the server. The headers in the negotiation response are the final word on how the data compression/decompression will be done between the client and server. Data compressed by the client must be decompressed by the server and vice versa. The client and server must adopt the same compression/decompression configuration parameters. We shall deal with this in detail in the future sections.
+
+The specification also discusses the DEFLATE algorithm. We adopt the [zlib](https://www.zlib.net) for doing raw compression and decompression. A pair of a compressor and a decompressor must be set up at both the ends of the connection. The server's decompressor decompresses messages compressed by the client's compressor and vice versa.
+
+### 3. An implementation of permessage-deflate based on swift-nio
+The [swift-nio](https://github.com/apple/swift-nio) framework provides an API which enables HTTP/WebSocket server implementations view the processing of data, read from or written to sockets, as a sequence of transformations that happen through a pipeline of handlers. An active connection is represented by a `Channel`. Data which is read from, or written to, a channel moves through a `ChannelPipeline` of inbound and outbound `ChannelHandlers`. An `EventLoop` is associated with every `Channel`. An `EventLoop`s is a thread-safe abstraction of a thread and provides features for asynchronous code execution using `EventLoopFutures` and `EventLoopPromises`.
```

Comment:
Can we reword the first bit of this sentence...?
"The [SwiftNIO](https://github.com/apple/swift-nio) framework provides an API which enables HTTP/WebSocket server implementations to view the processing of data, that has been read from or written to sockets, as...."